File: biocentral_api/dart/lib/src/clients/predict_client.dart
import 'package:biocentral_api/src/api.dart' as gen;
import 'package:biocentral_api/src/model/biocentral_prediction_model.dart';
import 'package:biocentral_api/src/model/model_metadata.dart';
import 'package:biocentral_api/src/model/prediction.dart';
import 'package:biocentral_api/src/model/prediction_request.dart';
import 'package:biocentral_api/src/model/task_dto.dart';
import 'package:biocentral_api/src/model/task_status.dart';
import 'package:built_collection/built_collection.dart';

import 'tasks/biocentral_server_task.dart';
import 'tasks/dto_handler.dart';
import 'tasks/submit_task.dart';

class _PredictDtoHandler extends DtoHandler<Map<String, List<Prediction>>> {
  @override
  Map<String, List<Prediction>>? handle(List<TaskDTO> dtos) {
    for (final dto in dtos) {
      if (dto.status == TaskStatus.FINISHED) {
        return dto.predictions?.toMap().map((k, v) => MapEntry(k, v.toList()));
      }
    }
    return null;
  }
}

class PredictClient {
  Future<List<ModelMetadata>?> getModelMetadata({
    required gen.BiocentralApi api,
  }) async {
    final predictionApi = api.getPredictionApi();
    final resp = await predictionApi.modelMetadataApiV1PredictionServiceModelMetadataGet();
    final metadataResponse = resp.data!;
    return metadataResponse.metadata.toList();
  }

  /// Start a prediction task using provided model names and sequences.
  Future<BiocentralServerTask<Map<String, List<Prediction>>>> predict({
    required gen.BiocentralApi api,
    required List<String> modelNames,
    required Map<String, String> sequenceData,
  }) async {
    final biocentralPredictionModelNames =
        BiocentralPredictionModel.values.where((bpm) => modelNames.contains(bpm.name));
    final predictionApi = api.getPredictionApi();
    final req = PredictionRequest((b) => b
      ..modelNames.replace(BuiltList<BiocentralPredictionModel>(biocentralPredictionModelNames))
      ..sequenceInput.replace(BuiltMap<String, String>(sequenceData)));

    final taskId =
        await submitTask(() => predictionApi.predictApiV1PredictionServicePredictPost(predictionRequest: req));
    final handler = _PredictDtoHandler();
    return BiocentralServerTask<Map<String, List<Prediction>>>(taskId: taskId, api: api, dtoHandler: handler);
  }
}

File: biocentral_api/python/biocentral_api/clients/predict_client.py
from tqdm.auto import tqdm
from typing import Dict, List

from .client_interface import ClientInterface
from .tasks import BiocentralServerTask, DTOHandler

from ..utils import calculate_sequence_hash
from .._generated import ApiClient, PredictionRequest, PredictionApi, TaskStatus, \
    TaskDTO, BiocentralPredictionModel, Prediction


class _PredictDTOHandler(DTOHandler):
    def __init__(self, hash2id: Dict[str, str]):
        self.hash2id = hash2id

    def handle_result(self, dtos: List[TaskDTO]):
        for dto in dtos:
            status = dto.status
            if status == TaskStatus.FINISHED:
                return dto.predictions

        return None

    def update_tqdm(self, dtos: List[TaskDTO], pbar: tqdm) -> tqdm:
        for dto in dtos:
            status = dto.status
            match status:
                case TaskStatus.PENDING:
                    pbar.set_description(f"Waiting for prediction task to start..")
                case TaskStatus.RUNNING:
                    pbar.set_description(f"Predicting..")
                case TaskStatus.FINISHED:
                    pbar.set_description(f"Finished predictions!")
                    break
                case TaskStatus.FAILED:
                    pbar.set_description(f"Predictions failed!")
                    break
        return pbar

    def get_tqdm_initial_description(self) -> str:
        return f"Running predictions for {len(self.hash2id)} sequence(s).."


class PredictClient(ClientInterface):
    def predict(self, api_client: ApiClient, model_names: List[BiocentralPredictionModel],
                sequence_data: Dict[str, str]) -> BiocentralServerTask[Dict[str, List[Prediction]]]:
        assert len(sequence_data) > 0, "No sequences provided"
        assert len(sequence_data.values()) == len(set(sequence_data.values())), "Duplicate sequences provided"

        hash2id = {calculate_sequence_hash(seq): seq_id for seq_id, seq in sequence_data.items()}

        prediction_request = PredictionRequest(model_names=model_names, sequence_input=sequence_data)
        api_instance = PredictionApi(api_client)
        task_id = self._submit_task(
            endpoint_caller=lambda: api_instance.predict_api_v1_prediction_service_predict_post(
                prediction_request))

        predict_dto_handler = _PredictDTOHandler(hash2id)
        biocentral_server_task = BiocentralServerTask(task_id=task_id,
                                                      api_client=api_client,
                                                      dto_handler=predict_dto_handler)
        return biocentral_server_task
